### DISCUSSION TITLE
Allow top-level functions outside classes

### DISCUSSION THREAD

**Discussion by Nefcanto** [dotnet/csharplang#8136] [Language Ideas]
Let us be able to create a standalone function outside a class or a namespace, like other languages, like JS or Python.

```
public static bool IsEven(int number)
{
    // logic
}
```

**Comment by KennethHoff** [comment]
How do you call this? Do you simply import the namespace and you're able to call the function?

**Comment by Nefcanto** [comment]
@KennethHoff, just the way people call global functions in other languages. They are available everywhere.

You can call it like `IsEven(8)` anywhere in your code. It's not inside a namespace. It's inside the global namespace.

And if people created two identical functions with the same name, then the code won't compile. But if they were different in signature, then both functions would be available as overloads.

**Comment by HaloFour** [comment]
See: #1180

**Comment by Nefcanto** [comment]
@HaloFour, that was a long thread. I did not read all of it. Just some comments at the beginning and some comments at the end. And after 7 years the C# team has done nothing related to that request. Is there a specific comment to redirect me to?

**Reply by ufcpp** [reply]
https://github.com/dotnet/csharplang/issues/3117#issuecomment-728290608
Likely Never

**Reply by Nefcanto** [reply]
@ufcpp, thank you. I saw the response and the reason for the rejected status.

**Reply by HaloFour** [reply]
I'll note that I do recall discussions on [Discord](https://aka.ms/dotnet-discord) more recently where a member of the language design team expressed interest in modules, so it's quite possible that the conversation isn't over.

**Reply by Nefcanto** [reply]
@HaloFour, the point is C# lacks some common and useful features like mixin, traits, top-level functions, and stuff like that. If the C# team and Microsoft don't change their mind, developers migrate. I remember back in the days when Microsoft share was dropping and they decided to go open-source. They changed their mind. They bought GitHub and open-sourced the entire .NET stack. This is the same story. If they don't make C# richer in terms of becoming more versatile, we might choose to put it aside as a team in our company.

**Reply by HaloFour** [reply]
Said language features need to demonstrate their value in terms of what problems they solve.  It's not sufficient to want to copy what other languages have done.

As demonstrated in detail in that other thread, top-level functions provide very little benefit to the language.  There's little/no difference between defining those methods top-level within a namespace vs. as static in a class.  The language already makes it possible to import those static methods across an entire project with a single line of code.  So in terms of production and consumption, it's practically a wash.  So, the question is what actual benefit comes from defining functions as top-level, aside a stylistic preference against using the word `class`.  That's ultimately what any conversation about "modules" would also have to answer.  Otherwise, in terms of ROI, it's simply not worth it.

**Reply by Nefcanto** [reply]
@HaloFour, you're right. Any feature should prove its worth first. But based on my experience I know that when a language team creates a couple of workarounds to mimic other languages, then it means that it's better to implement those features in real. Instead of having `using static SomeClassName` which is a workaround, why not have something built in the programming language?

Having top-level functions is worthwhile. Of course, it was worthy enough for the team to create a pseudo way so that teams would be able to have it (to import it). Yet those solutions are just some workarounds. They are not real implementations. They are not first-class citizens in that language.

I argue that most of the time (not always) the community knows the best because of the collective wisdom. JSX (HTML-in-JS) proved to be successful, and the community chose it over JS-in-HTML model of Angular. The community voted for it (200K stars on React).

Now when you look at the popularity of other languages and frameworks, you clearly see the trace of mixin, and traits (composability). I'm sure we will see them enter C# in the future. They have no choice but to add it.

**Reply by CyrusNajmabadi** [reply]
> Yet those solutions are just some workarounds. They are not real implementations. They are not first-class citizens in that language.

What makes it not a real implementation? What makes it not first class? 

What would the difference be?

**Reply by Nefcanto** [reply]
@CyrusNajmabadi, this list:

1. Not intuitive. A novice developer would never guess it.
2. Still needs a namespace and a static class as wrappers (boilerplate)
3. Still needs a `using` boilerplate

**Reply by HaloFour** [reply]
> They are not real implementations. They are not first-class citizens in that language.

They solve the problem, through "first class" mechanism that also address other problems.

> Now when you look at the popularity of other languages and frameworks, you clearly see the trace of mixin, and traits (composability). I'm sure we will see them enter C# in the future. They have no choice but to add it.

Which languages would that be?  Given the languages that C# already shares the top 5 list with, I seriously doubt that assertion.  Those features would definitely be nice to have, but C# isn't going to chase them for the sake of being "popular", especially given that C# is already more popular than most of those other languages.  C# will add features because they solve problems.

> A novice developer would never guess it.

A novice developer will have to learn how features work in any language.  Once they do, they will understand how do use said feature.

A novice developer who switches between two programming languages will realize that the workflows that they learned to solve problems in one language don't necessarily apply to the other language.  They will need to learn new idioms and new ways to approach the problem.  Attempting to write one language like another is a quick way to run into friction.  This is true of any two languages. 

The argument that C# must do this or die will not resonate.  C# will be fine without it.

**Reply by siegfriedpammer** [reply]
> Still needs a namespace

You can put the class in the global namespace.

Honestly, I dislike the idea of having global symbols that pollute the global namespace, i. e., are available without imports.

Ever had to use a library that defines extension methods on System.Object? It's the same kind of pollution and it's really bad.

**Reply by HaloFour** [reply]
> Still needs a namespace and a static class as wrappers (boilerplate)

What you see as boilerplate is also what solves so many of the design problems around top-level functions.

If you read through the other thread you'll note that there are a lot of different opinions as to how top-level functions should work, regarding their organization, discoverability and visibility.  By having the developer explicitly declare the "module" as a static class, you solve for all of those problems.  The flexibility of `using static` and `global using` enables any import strategy desired, from one big module imported once across the entire project to several modules imported granularly where desired.  This is particularly important if there is any desire to export the functions to be imported by other assemblies, where the public surface matters a great deal and most .NET languages have no concept of functions that exist outside of classes.

Any implementation of top-level functions would have to solve for all of these design problems.  It will either need very similar "boilerplate", or it would end up having to enforce some default opinions that aren't likely to align with your own.

**Reply by Nefcanto** [reply]
@HaloFour, we're talking abstract here. The composition is good. Your refrigerator is composed. Your car is composed. And boilerplate is bad. So, a good language is a language that enables composition, with less boilerplate. This is a true sentence logically.

That's what JSX did. More composability with less verbosity. Now compare JSX with ASP.NET Razor components. JSX easily wins and ASP.NET Razor Components easily loses. Not because I say so. Stats show that. I have worked in multiple teams, and multiple times for multiple projects we failed to make ASP.NET Razor Components work for us and we managed to make React work for us. Why? More composability, with less verbosity. That's why React team went from class components to function components. Please let me repeat it one more time. More composability with less verbosity.

Yes, C# is one of the top-file languages, because it managed to add features to attract the community. But had it stayed in its position years ago, it would have been dead by now.

And learning curve also matters. Any developer can create a function in any language in less than 10 seconds. But to call that function from another file, that's where boilerplate and complexity show up.

I love C#. I've been a C# developer for 18 years. But as things get more complicated and more dynamic, I feel the need for composition (mixin or traits), and the need for less verbosity.

**Reply by CyrusNajmabadi** [reply]
> and the need for less verbosity.

You can get that today.  The language provides the building blocks to make this possible, while also avoiding the pitfalls and problems with top level functions elsewhere. 

It's still unclear why that is not suitable here. 

> this list:

> Not intuitive. A novice developer would never guess it.

Nothing is intuitive for a novice. Everything must be learned. 

> Still needs a namespace and a static class as wrappers (boilerplate)

It doesn't need a namespace. Though we'd definitely recommend using a namespace to avoid polluting the global namespace. 

> Still needs a using boilerplate

That's one line written once in the entire project. If that's too onerous, then I'm not sure we can find common ground here. We're not going to continually invest into lots of work to make it so you can avoid writing a single statement.

**Reply by Nefcanto** [reply]
@CyrusNajmabadi, to be fair, yes you're right that C# is less verbose these days. 

1. We enjoy `global using` to reduce our repetitive `using System;` and stuff like that.
2. We enjoy de-indenting our entire codebase one tab backwards because of `namespace X;`, 
3. we enjoy not indenting our code for disposable `using` statements. 
4. We do enjoy being able to create arrays like JS.
5. We love the spread syntax.

But is that trend not apparent already? The C# team has already started becoming less verbose. Why not continuing this path?

> Nothing is intuitive for a novice. Everything must be learned


I'm not denying the path to learning. I'm emphasizing a reduction in the learning curve. I talk about the experience. Hire 10 novice developers. Teach them C# and teach them JS. They grasp JS much easier. If I'm wrong, please correct me. I have done it many times over and over, since I'm in position of recruitment of interns to educate them for companies. And JS is a fairly complex language if you dive deep into it. Yet it's also easy for beginners. Because novice developers understand a function very easily. It's in base math. But as soon as you mix it with **write a class first then write that function** it becomes 10-fold harder. Why should it be like that?

**Reply by HaloFour** [reply]
> But as soon as you mix it with write a class first then write that function it becomes 10-fold harder. Why should it be like that?
 
https://learn.microsoft.com/en-us/dotnet/csharp/fundamentals/program-structure/top-level-statements

**Reply by CyrusNajmabadi** [reply]
> But is that trend not apparent already?

We have created features when there isn't something existing to solve that issue. That doesn't apply here.

**Reply by CyrusNajmabadi** [reply]
> If I'm wrong, please correct me

That's not what we have seen. Js has so much more in terms of having to guide people away from all the problems in it.

**Reply by CyrusNajmabadi** [reply]
> But as soon as you mix it with write a class first then write that function it becomes 10-fold harder.

The more concepts you have, the harder it gets. So teaching them one thing and how you can use it for all these purposes is easier than having them to teach a bunch of disparate concepts.

**Comment by colejohnson66** [comment]
I can understand your plight and desire for this feature, and it’s one that, when I started learning C# (about 15 years ago), I used to wish for myself.

However, @HaloFour is right: this is something that can now be done with existing language features. Specifically, `using static`: `using static MyNamespace.MyGlobalClass`. Combine that with `partial` classes and you can spread the methods over as many files as you wish.

If you want to "pollute" the global namespace, `global using` exists: `global using static MyNamespace.MyGlobalNamespace`.
